lib/Mojo/Redis/Connection.pm

```diff
@@ -163,22 +168,26 @@ sub _on_close_cb {
   };
 }
 
-sub _on_read_cb {
-  my $self = shift;
+sub _on_message_cb {
+  my $self     = shift;
+  my $protocol = $self->protocol;
 
   Scalar::Util::weaken($self);
-  return sub {
-    return unless $self;
-    my ($stream, $chunk) = @_;
-    do { local $_ = $chunk; s!\r\n!\\r\\n!g; warn "[@{[$self->_id]}] >>> ($_)\n" } if DEBUG;
-    $self->protocol->parse($chunk);
-  };
-}
 
-sub _parse_message_cb {
-  my $self = shift;
+  if ($protocol->can('encoding')) {
+    return sub {
+      my ($protocol, $msg) = @_;
+      my $p = shift @{$self->{waiting} || []};
+
+      if ($msg->{type} eq '-') {
```

Comment:
This won't emit an error if there is an error contained within a multi-bulk reply somewhere; it seems like the current unwrapping method does this.

Reply:
I'm not sure the best way to handle it, should a sub-error (like in ->exec response) cause the whole command to error? But we have no way to differentiate errors within the returned arrayref from strings.

Reply:
How does this work for array types? Doesn't it need to change the data from hashrefs to their contents (and so on if it contains an array...)

Reply:
I'm not sure how to write a test for error inside arrays. Do you know how to provoke such an error? If so, then I think I want to return it as a single error, since that's what I support in Connection.pm: https://github.com/jhthorsen/mojo-redis/blob/3901fef30999f7eabea5806d6fac7e86f4d9f489/lib/Mojo/Redis/Connection.pm#L216